Mathlib/Algebra/Lie/Derivation/Basic.lean

```diff
@@ -0,0 +1,349 @@
+- The set of Lie derivations from a Lie algebra to itself is a Lie algebra.
+
+## Implementation notes
+
+- Mathematically, a Lie derivation is just a derivation on a Lie algebra. However, the current
+implementation of `Derivation` requires a commutative associative algebra, so is incompatible
+with the setting of Lie algebras. Initially, this file is a copy-pasted adaptation of the
+`RingTheory.Derivation.Basic` file.
+- Since we don't have right actions of Lie algebras, the second term in the Leibniz rule is written
+as `- [b, D a]`. Within Lie algebras, skew symmetry restores the expected definition `[D a, b]`.
+-/
+
+/-- A Lie derivation `D` from the Lie `R`-algebra `L` to the `L`-module `M` is an `R`-linear map
+that satisfies the Leibniz rule `D [a, b] = [a, D b] - [b, D a]`. -/
+structure LieDerivation (R L M : Type*) [CommRing R] [LieRing L] [LieAlgebra R L]
+    [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+    extends L →ₗ[R] M where
+  protected leibniz' (a b : L) : toLinearMap ⁅a, b⁆ = ⁅a, toLinearMap b⁆ - ⁅b, toLinearMap a⁆
+
+open scoped BigOperators
+
+/-- The `LinearMap` underlying a `LieDerivation`. -/
+add_decl_doc LieDerivation.toLinearMap
+
+namespace LieDerivation
+
+section
+
+variable {R L M : Type*} [CommRing R] [LieRing L] [LieAlgebra R L]
+    [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+
+variable (D : LieDerivation R L M) {D1 D2 : LieDerivation R L M} (r : R) (a b : L)
+
+instance : FunLike (LieDerivation R L M) L M where
+  coe D := D.toFun
+  coe_injective' D1 D2 h := by cases D1; cases D2; congr; exact DFunLike.coe_injective h
+
+instance : AddMonoidHomClass (LieDerivation R L M) L M where
+  map_add D := D.toLinearMap.map_add'
+  map_zero D := D.toLinearMap.map_zero
+
+-- Not a simp lemma because it can be proved via `coeFn_coe` + `toLinearMap_eq_coe`
+theorem toFun_eq_coe : D.toFun = ⇑D :=
+  rfl
+
+/-- See Note [custom simps projection] -/
+def Simps.apply (D : LieDerivation R L M) : L → M := D
+
+initialize_simps_projections LieDerivation (toFun → apply)
+
+attribute [coe] toLinearMap
+
+instance hasCoeToLinearMap : Coe (LieDerivation R L M) (L →ₗ[R] M) :=
+  ⟨fun D => D.toLinearMap⟩
+
+#noalign derivation.to_linear_map_eq_coe -- Porting note: not needed anymore
+
+@[simp]
+theorem mk_coe (f : L →ₗ[R] M) (h₁) : ((⟨f, h₁⟩ : LieDerivation R L M) : L → M) = f :=
+  rfl
+
+@[simp, norm_cast]
+theorem coeFn_coe (f : LieDerivation R L M) : ⇑(f : L →ₗ[R] M) = f :=
+  rfl
+
+theorem coe_injective : @Function.Injective (LieDerivation R L M) (L → M) DFunLike.coe :=
+  DFunLike.coe_injective
+
+@[ext]
+theorem ext (H : ∀ a, D1 a = D2 a) : D1 = D2 :=
+  DFunLike.ext _ _ H
+
+theorem congr_fun (h : D1 = D2) (a : L) : D1 a = D2 a :=
+  DFunLike.congr_fun h a
+
+protected theorem map_add : D (a + b) = D a + D b :=
+  map_add D a b
+
+protected theorem map_zero : D 0 = 0 :=
+  map_zero D
+
+@[simp]
+theorem map_smul : D (r • a) = r • D a :=
+  D.toLinearMap.map_smul r a
```

Comment:
These are all available via the general lemmas which require only a `LinearMapClass` instance
```suggestion
```

Reply:
(indeed one of the main reasons for these `LinearMapClass` typeclasses is to save duplicating these lemmas everywhere)